### PR TITLE
tcpstats for OS X

### PR DESCRIPTION
**Pull request by girishmotwani**

### CODE REVIEW

File: plt_darwin.go [hunk 1]
```diff
@@ -43,10 +48,27 @@ func getNetDevStats(stats *ethrNetStat) {
 }
 
 func getTcpStats(stats *ethrNetStat) {
+	// use netstat to get the interface stats
+	args := []string{"-s", "-p", "tcp"}
+	output, err := exec.Command(netstatPath, args...).Output()
```
[the remark this answers is not in the document]
Comment (girishmotwani) [reply plt_darwin.go]: One could look at what netstat does and call the underlying sysctl "net.inet.tcp.stats" to fetch this. That would mean looking up the OS X definition for struct tcpstat and then sticking that in Ethr.  The netstat tool is included in the distribution and does all the hard work to get the stats.

File: utils.go [hunk 2]
```diff
@@ -235,21 +236,21 @@ func roundUpToZero(n int64) int64 {
...
 func getFd(conn net.Conn) uintptr {
-	var fd uintptr
-	var rc syscall.RawConn
-	var err error
-	switch ct := conn.(type) {
-	case *net.TCPConn:
-		rc, err = ct.SyscallConn()
-		if err != nil {
-			return 0
+	switch conn.(type) {
```
[the remark this answers is not in the document]
Comment (girishmotwani) [reply utils.go]: I saw the following errors in compilation:

mbp15:~ go$ go build github.com/Microsoft/Ethr
# github.com/Microsoft/Ethr
src/github.com/Microsoft/Ethr/utils.go:224: undefined: syscall.RawConn
src/github.com/Microsoft/Ethr/utils.go:228: ct.SyscallConn undefined (type *net.TCPConn has no field or method SyscallConn)

[the remark this answers is not in the document]
Comment (pgmsft) [reply utils.go]: Which golang version you are using? @fsouza didn't hit this issue so wondering if this is version specific.

File: plt_darwin.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (pgmsft) [reply plt_darwin.go]: Can we use the API as it is much lighter weight and easier to parse, less susceptible to breaking than netstat output parsing?

File: utils.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (girishmotwani) [reply utils.go]: 1.8

[the remark this answers is not in the document]
Comment (pgmsft) [reply utils.go]: Can you try with 1.10+? That is the supported version.

**Comment by pgmsft** [comment]
I think we can do something like below to make it work:

In getTcpStat function (pseudo implementation shown below) we can use "net.inet.tcp.stats" for name. This would generate tcpstat output. The definition of tcpstat is here: https://github.com/apple/darwin-xnu/blob/master/bsd/netinet/tcp_var.h

Refer: https://opensource.apple.com/source/network_cmds/network_cmds-85/netstat.tproj/inet.c

func getTcpStat() {
	val, err := syscall.Sysctl(name)
	if err != nil {
		return err
	}

	buf := []byte(val)

	bbuf := bytes.NewBuffer([]byte(val))
	return binary.Read(bbuf, binary.LittleEndian, data)
}

I don't have a MacOS to try this myself, but if someone else can take a shot, then we can validate if this works.

**Comment by girishmotwani** [comment]
@pgmsft  I'll update the patch shortly with the sysctl call

**Pull request merged**